arraycontext/impl/pytato/__init__.py

```diff
@@ -232,77 +251,89 @@ def __init__(self, queue: "cl.CommandQueue", allocator=None,
+        return rec_map_array_container(_wrapper, array)
 
-        entrypoint = program.default_entrypoint.name
+    # {{{ ArrayContext interface
 
-        # {{{ preprocess args
+    def zeros_like(self, ary):
+        def _zeros_like(array):
+            return self.zeros(array.shape, array.dtype)
 
-        processed_kwargs = {}
+        return self._map_container(_zeros_like, ary)
 
-        for kw, arg in sorted(kwargs.items()):
-            if isinstance(arg, (pt.Array,) + SCALAR_CLASSES):
-                pass
-            elif isinstance(arg, TaggableCLArray):
-                arg = self.thaw(arg)
-            else:
-                raise ValueError(f"call_loopy argument '{kw}' expected to be an"
-                                 " instance of 'pytato.Array', 'Number' or"
-                                 f"'TaggableCLArray', got '{type(arg)}'")
+    def from_numpy(self, array):
+        import pytato as pt
+        import arraycontext.impl.pyopencl.taggable_cl_array as tga
 
-            processed_kwargs[kw] = arg
+        def _from_numpy(ary):
+            return pt.make_data_wrapper(
+                tga.to_device(self.queue, ary, allocator=self.allocator)
+                )
 
-        # }}}
+        return with_array_context(
+            self._map_container(_from_numpy, array, (np.ndarray,), strict=True),
+            actx=self)
 
-        return call_loopy(program, processed_kwargs, entrypoint)
+    def to_numpy(self, array):
+        def _to_numpy(ary):
+            return ary.get(queue=self.queue)
+
+        return with_array_context(
+            self._map_container(_to_numpy, self.freeze(array)),
+            actx=None)
 
     def freeze(self, array):
+        if np.isscalar(array):
+            return array
+
         import pytato as pt
         import pyopencl.array as cla
-        import loopy as lp
 
-        from arraycontext.container import ArrayT
         from arraycontext.container.traversal import rec_keyed_map_array_container
         from arraycontext.impl.pytato.utils import (_normalize_pt_expr,
                                                     get_cl_axes_from_pt_axes)
         from arraycontext.impl.pyopencl.taggable_cl_array import (to_tagged_cl_array,
                                                                   TaggableCLArray)
         from arraycontext.impl.pytato.compile import _ary_container_key_stringifier
 
-        array_as_dict: Dict[str, Union[cla.Array, TaggableCLArray,
-                                       pt.Array]] = {}
+        array_as_dict: Dict[str, Union[cla.Array, TaggableCLArray, pt.Array]] = {}
         key_to_frozen_subary: Dict[str, TaggableCLArray] = {}
         key_to_pt_arrays: Dict[str, pt.Array] = {}
 
-        def _record_leaf_ary_in_dict(key: Tuple[Any, ...],
-                                     ary: ArrayT):
+        def _record_leaf_ary_in_dict(
+                key: Tuple[Any, ...],
+                ary: Union[cla.Array, TaggableCLArray, pt.Array]):
```

Comment:
` -> None`?

Reply:
It returned the `ary`, but it should work just fine without returning anything, so changed in https://github.com/inducer/arraycontext/compare/f4ab671ce9eb96e9dd2397d9a77d5e7ac7c9be36..971df793aed6b5b107be09588c0791a2f664c2b9